File: BOJ13023_ABCDE.cpp
#include <iostream>
#include <vector>
using namespace std;
typedef uint16_t human;
typedef vector<human> relationlist;
vector<relationlist> mylist;
bool fivebridges(human A){
    for (human B : mylist[A]) {
        for (human C : mylist[B]) {
            if (C==A) continue;
            for (human D : mylist[C]) {
                if (D==A) continue;
                if (D==B) continue;
                for (human E : mylist[D]) {
                    if (E==A) continue;
                    if (E==B) continue;
                    if (E==C) continue;
                    return true;
                }
            }
        }
    }
    return false;
}
int main() {
    size_t N, M, rfrom, rto;
    cin >> N >> M;
    mylist.resize(N);

    while (M--) {
        cin >> rfrom >> rto;
        mylist[rfrom].push_back(rto);
        mylist[rto].push_back(rfrom);
    }

    for (human i = 0; i < N; ++i){
        if (fivebridges(i)) {
            cout << '1' << endl;
            return 0;
        }
    }
    cout << '0' << endl;
}

File: BOJ13023_ABCDE.java
import java.util.Scanner;
import java.util.ArrayList;
@SuppressWarnings("unchecked")
public class BOJ13023_ABCDE {
    static ArrayList<Integer>[] mylist;
    static boolean fivebridges(int A){
        for (int B : mylist[A]) {
            for (int C : mylist[B]) {
                if (C==A) continue;
                for (int D : mylist[C]) {
                    if (D==A) continue;
                    if (D==B) continue;
                    for (int E : mylist[D]) {
                        if (E==A) continue;
                        if (E==B) continue;
                        if (E==C) continue;
                        return true;
                    }
                }
            }
        }
        return false;
    }
    public static void main(String args[]) {
        Scanner in = new Scanner(System.in);
        int N, M, rfrom, rto;
        N = in.nextInt();
        M = in.nextInt();
        mylist = new ArrayList[N];
        for (int i = 0; i < N; i++) mylist[i] = new ArrayList<Integer>();
        
        while (M-- != 0) {
            rfrom = in.nextInt();
            rto = in.nextInt();
            mylist[rfrom].add(rto);
            mylist[rto].add(rfrom);
        }
        in.close();

        for (int i = 0; i < N; ++i){
            if (fivebridges(i)) {
                System.out.println(1);
                return;
            }
        }
        System.out.println(0);
    }
}